### PR TITLE
[DNNL][INT8][FP32] MatMul

### PR DESCRIPTION
**Pull request by Sand3r-**
This PR introduces:
* FP32 and INT8 implementations of DNNL MatMul operator
* UTs to test functionality

### CODE REVIEW

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: I would just suggest that you could consider renaming dim_x -> mat_dim_x like in matmul_op.cc InferShape https://github.com/PaddlePaddle/Paddle/pull/23395/files#diff-126803a76a1b0cbf6198f30e23058f8dR336.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (sfraczek) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: This one stands out in the middle. What do you think about for example moving it to bottom of block or something?

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (lidanqing-vv) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: skip_check_grad_ci is recommended in such case. Like following
https://github.com/PaddlePaddle/Paddle/pull/21985/files

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Done.

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Done.

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (Sand3r-) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: Done.

**Comment by lidanqing-vv** [comment]
@luotao1  Hi, please give an approval about the UT `@skip_check_grad_ci`. This is blocking the CI to pass. Thanks

**Comment by Sand3r-** [comment]
@luotao1 

PR-CI-Coverage failed because it cannot find error_codes when building shell.cc. I didn't touch that file. What should I do?
```
[2020-04-03 21:24:05]  [  3%] Linking CXX static library libfs.a
[2020-04-03 21:24:05]  In file included from /paddle/paddle/fluid/platform/enforce.h:41:0,
[2020-04-03 21:24:05]                   from /paddle/paddle/fluid/framework/io/shell.cc:16:
[2020-04-03 21:24:05]  /paddle/paddle/fluid/platform/errors.h:23:50: fatal error: paddle/fluid/platform/error_codes.pb.h: No such file or directory
[2020-04-03 21:24:05]  #include "paddle/fluid/platform/error_codes.pb.h"
[2020-04-03 21:24:05]                                                    ^
[2020-04-03 21:24:05]  compilation terminated.
```

**Comment by luotao1** [comment]
The fixed PR is #23467, please rerun after it is merged.

**Comment by sfraczek** [comment]
Can you add matmul to `--quantized_ops` in `python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py`  and   `QAT2_NLP_QUANTIZED_OPS` in `python/paddle/fluid/contrib/slim/tests/CMakeLists.txt`?

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (wojtuss) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: The name is a bit misleading, as the function is about the shape, not about the matrix.
Same below

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: Aren't the dimensions too small for CI? I remember a requirement that tensors should have at least 100 elements.

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (wojtuss) [inline python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: You can drop the grad checking methods after adding `@skip_check_grad_ci`.

**Review by wojtuss — COMMENTED** [review]
I find this PR very concise and elegant. My comments are very minor and, as the CI is not stable and the queue is very long, please, do not consider them as requests for changes. They can be addressed later in a refactoring PR if necessary.

**Comment by Sand3r-** [comment]
> Can you add matmul to `--quantized_ops` in `python/paddle/fluid/contrib/slim/quantization/quantization_mkldnn_pass.py` and `QAT2_NLP_QUANTIZED_OPS` in `python/paddle/fluid/contrib/slim/tests/CMakeLists.txt`?

This PR doesn't introduce any changes to cpu quantise pass, so that wouldn't work, nope

File: python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py
Comment (lidanqing-vv) [reply python/paddle/fluid/tests/unittests/mkldnn/test_matmul_mkldnn_op.py]: @Sand3r-  https://github.com/PaddlePaddle/Paddle/wiki/OP-Test-Input-Shape-Requirements-(English-Version) The requirements wojtek mentioned could be found here. @skip_check_grad is here too

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Sand3r-) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: What other name do you suggest instead? I think it conveys the meaning that it accepts a vector of size N and converts it into a matrix of size 1 x N. Similarly for the other function.

Comment (wojtuss) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: Maybe `RowMatrixDimsFromVector`, or `VectorDims2RowMatrixDims`?
It is only a suggestion.

Comment (sfraczek) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: But it's the same function as already exists here:
https://github.com/PaddlePaddle/Paddle/pull/23395/files/471e894b65b0cf2db48de2ee2b0c330d3cbd7815#diff-126803a76a1b0cbf6198f30e23058f8dL39

**Review by wojtuss — APPROVED** [review]
LGTM

**Comment by Sand3r-** [comment]
@luotao1 
All builds have passed (apart from the one that needs approval) and internal review is already finished. Could you please start your review?

**Review by sfraczek — APPROVED** [review]
LGTM

File: paddle/fluid/operators/matmul_op.cc
Comment (luotao1) [inline paddle/fluid/operators/matmul_op.cc:470]: Are these attributes all needed?

Comment (Sand3r-) [reply paddle/fluid/operators/matmul_op.cc:470]: Yes, all int8 compute operators need these attributes. 
Conv2d: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/conv_op.cc#L307-L326
FC: https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/fc_op.cc#L145-L165

**Review by luotao1 — APPROVED** [review]
LGTM

**Comment by lidanqing-vv** [comment]
@luotao1  Can you please help restart PR_CI_CPU_Py2 (Compile)? As you already approved, but this CI was not restarted.

**Pull request merged**
